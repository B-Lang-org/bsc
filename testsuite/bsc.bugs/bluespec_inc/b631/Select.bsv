import List::*;

interface Property#(parameter type a);
  method ActionValue#(a) advance(a d);
endinterface

module mkTestProp (Property#(int));

  method ActionValue#(int) advance (int data);
    return 0;
  endmethod
endmodule


interface Assertion#(parameter type a);
  method Action advance(a d);
endinterface


module mkAssert#(Integer n) (Assertion#(int));

  List#(Property#(int)) ps <- replicateM(n, mkTestProp);
  int len = fromInteger(n);

  method Action advance (int data);
  action
    int pres = 1;
    //for (Integer k = 0; fromInteger(k) < len; k = k + 1) //This works
    for (Integer k = 0; fromInteger(k) <= len; k = k + 1) //This doesn't work
    begin
      pres <- (ps[k]).advance(data);
    end
  endaction
  endmethod

endmodule

(* synthesize *)
module mkTest ( Empty );

  Reg#(int) x <- mkReg(0);

  function Bool lambda(int y);
    if ((y & 1) == 1) return True; else return False;
  endfunction

  Assertion#(int) as <- mkAssert(5);

  rule foo (True);
   x <= x + 1;
   as.advance(x);
  endrule
endmodule
