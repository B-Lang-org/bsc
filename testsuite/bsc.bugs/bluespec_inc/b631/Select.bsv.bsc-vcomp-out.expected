checking package dependencies
compiling Select.bsv
code generation for mkTest starts
Error: "Select.bsv", line 31, column 18: (S0015)
  Bluespec evaluation-time error: index 5 out-of-range (list selection)
  During elaboration of the body of rule `foo' at "Select.bsv", line 49,
  column 8.
  During elaboration of `mkTest' at "Select.bsv", line 39, column 8.
